src/Components/Endpoints/src/Discovery/PageCollectionBuilder.cs

```diff
@@ -0,0 +1,153 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// Represents the list of pages in a <see cref="ComponentApplicationBuilder"/>.
+/// </summary>
+internal class PageCollectionBuilder
+{
+    private readonly List<PageComponentBuilder> _pages = new();
+
+    internal void Combine(PageCollectionBuilder pages)
+    {
+        for (var i = 0; i < pages._pages.Count; i++)
+        {
+            var pageToAdd = pages._pages[i];
+            var found = false;
+            for (var j = _pages.Count - 1; j > 0; j--)
+            {
+                var page = _pages[j];
+                if (page.Equals(pageToAdd))
+                {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found)
+            {
+                _pages.Add(pageToAdd);
+            }
+        }
+    }
+
+    internal void Exclude(PageCollectionBuilder pages)
+    {
+        for (var i = 0; i < pages._pages.Count; i++)
+        {
+            var pageToRemove = pages._pages[i];
+            for (var j = _pages.Count - 1; j > 0; j--)
+            {
+                var page = _pages[j];
+                if (page.Equals(pageToRemove))
+                {
+                    _pages.RemoveAt(j);
+                    break;
+                }
+            }
+        }
+    }
+
+    internal void RemoveFromAssembly(string name)
+    {
+        for (var i = _pages.Count - 1; i > 0; i--)
+        {
+            if (_pages[i].HasSource(name))
+            {
+                _pages.RemoveAt(i);
+            }
+        }
+    }
+
+    internal void AddFromLibraryInfo(IEnumerable<PageComponentBuilder> pages)
+    {
+        _pages.AddRange(pages);
+    }
+
+    internal PageComponentInfo[] ToPageCollection()
+    {
+        if (_pages.Count == 0)
+        {
+            return Array.Empty<PageComponentInfo>();
+        }
+
+        var list = new List<PageComponentInfo>();
+        // Reuse a buffer for computing the metadata
```

Comment:
Given the immense amount of allocations going on during app startup (and even specifically during this discover process), it's unclear that the micro-optimization of not just using a `List<object>` here is beneficial, especially if we need to deal with resizing the array if it overflows etc. Am I missing something about which this particular piece benefits from this optimization?